### PR TITLE
scripts: support compile_commands.json in gen_app_partitions.py

### LINKED ISSUE #40590: gen_app_partitions scans object files unrelated to current image
**Describe the bug**
The `gen_app_partitions.py` does a directory walk here:
https://github.com/zephyrproject-rtos/zephyr/blob/f2d2a48b96172c768ec81ed1f1962ece6eef6b87/scripts/gen_app_partitions.py#L143-L149

This is wrong for two reasons:
- It will include left-over objects that is no longer part of the build, if sample has been reconfigured
  (incremental builds will not clean up)
- When external project is used, and the binary dir of that project is a sub-folder of the current sample, then the object files of the external image will also be processed, although they are completely unrelated to current image build.

**To Reproduce**
tricky, needs special skills.
Ask @tejlmand.

**Expected behavior**
Only object files for current image should be considered.

**Impact**
What impact does this issue have on your progress (e.g., annoyance, showstopper)

**Logs and console output**
If applicable, add console logs or other types of debug information
e.g Wireshark capture or Logic analyzer capture (upload in zip archive).
copy-and-paste text and put a code fence (\`\`\`) before and after, to help
explain the issue. (if unable to obtain text log, add a screenshot)

**Environment (please complete the following information):**
 - OS: Linux, but same is expected on other OS'es
 - Toolchain Zephyr SDK
 - f2d2a48b96172c768ec81ed1f1962ece6eef6b87

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by tejlmand**
Fixes: #40590

This commit updates gen_app_partitions.py to include only files present
in the current build by extracting the information from the CMake
generated `compile_commands.json` file.

This ensures that object files in sub-projects, such as `empty_cpu0`,
will not be considered by the script.

Using the compile_commands.json instead of walking the whole build tree
for finding object files also improves performance:

Time of executing `gen_app_partitions.py` (Old):
__________________________
Executed in  480.06 millis
   usr time  425.83 millis
   sys time   49.55 millis

Time of executing `gen_app_partitions.py` (New):
________________________________________________________
Executed in   76.22 millis
   usr time   49.00 millis
   sys time   24.59 millis

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

### CODE REVIEW

**Comment by dcpleung** [comment]
Some concerns:
* This requires all supported toolchains to use `-o` to indicate output file. Visual Studio uses `/` instead of `-` for command line arguments (though we don't support it currently). I think all of our currently supported toolchains uses `-o`, so this may not be an issue.
* This only works for compiled output. If someone decides to include a prebuilt library, this will skip the partitions defined in that library.

File: cmake/app/boilerplate.cmake [hunk 1]
```diff
@@ -558,6 +558,14 @@ set(SOC_SERIES ${CONFIG_SOC_SERIES})
...
 set(SOC_FAMILY ${CONFIG_SOC_FAMILY})
 
+# For the gen_app_partitions.py to work correctly, we must ensure that
+# all targets exports their compile commands to fetch object files.
+# We enable it unconditionally, as this is also useful for several IDEs
+set(CMAKE_EXPORT_COMPILE_COMMANDS TRUE CACHE BOOL
+    "Export CMake compile commands. Used by gen_app_partitions.py script"
+    FORCE
+)
+
```
Comment (SebastianBoe) [inline cmake/app/boilerplate.cmake:568]: Checked the overhead for this.

Was unable to measure a time overhead, and the file generated was 244k.

So generating this unconditionally is OK.

Comment (SebastianBoe) [reply cmake/app/boilerplate.cmake:568]: Looks like we can now get rid of some code in twister.

/home/sebo/ncs/zephyr/scripts/pylib/twister/twisterlib.py
2045:            cmake_args.append("-DCMAKE_EXPORT_COMPILE_COMMANDS=1")

Comment (tejlmand) [reply cmake/app/boilerplate.cmake:568]: > Was unable to measure a time overhead, and the file generated was 244k.
>
> So generating this unconditionally is OK.

Thanks for checking.
Initial version was actually doing this conditionally, but feedback on Discord suggested to do this unconditionally.

So great you checked the impact.

**Comment by tejlmand** [comment]
> - This requires all supported toolchains to use -o to indicate output file. Visual Studio uses / instead of - for command line arguments (though we don't support it currently). I think all of our currently supported toolchains uses -o, so this may not be an issue.

Not to mention, that the snippet generated is `ld` based, meaning it's supported by GNU ld, and thus only toolchains that understand that format. afaik Visual Code built-in linker doesn't support ld files, so really doesn't matter that Visual Studio uses `/` instead of `-`.

> This only works for compiled output. If someone decides to include a prebuilt library, this will skip the partitions defined in that library.

Exactly as current solution. Only object files are considered, not prebuilt libraries.
Notice this in existing solution:
https://github.com/zephyrproject-rtos/zephyr/blob/f35cdfc1e4d3a59541bffcaefea739def7cd5468/scripts/gen_app_partitions.py#L143-L145

Not to mention that `ELFFile()` only supports elf files, that is either final elf or object files, not library archives.

**Review by dcpleung — APPROVED** [review]

File: cmake/app/boilerplate.cmake (hunk 1 above)
Comment (tejlmand) [reply cmake/app/boilerplate.cmake:568]: Done, `CMAKE_EXPORT_COMPILE_COMMANDS` removed in twister.

File: scripts/gen_app_partitions.py [hunk 2]
```diff
@@ -237,6 +260,8 @@ def main():
 
     if args.directory is not None:
         parse_obj_files(partitions)
```
Comment (SebastianBoe) [inline scripts/gen_app_partitions.py:262]: Do we need to retain the parse_obj_files function?

File: scripts/gen_app_partitions.py [hunk 3]
```diff
@@ -149,6 +150,26 @@ def parse_obj_files(partitions):
...
+    # Thereafter process each object file to find partitions
+    object_pattern = re.compile(r'-o\s+(\S*)')
+    with open(args.compile_commands_file, 'rb') as f:
+        commands = json.load(f)
+        for command in commands:
+            build_dir = command.get('directory')
+            compile_command = command.get('command')
+            compile_arg = object_pattern.search(compile_command)
+            obj_file = None if compile_arg is None else compile_arg.group(1)
+            if obj_file is not None:
```
Comment (SebastianBoe) [inline scripts/gen_app_partitions.py:164]: IIRC

if obj_file:

is more pythonic

**Review by SebastianBoe — APPROVED** [review]

File: scripts/gen_app_partitions.py (hunk 2 above)
Comment (tejlmand) [reply scripts/gen_app_partitions.py:262]: Not for Zephyr use-case afaik, but i'm unsure if anyone downstream is using this script also, so I didn't wanted to remove it straight away.

File: scripts/gen_app_partitions.py (hunk 3 above)
Comment (tejlmand) [reply scripts/gen_app_partitions.py]: fixed.

**Review by nashif — APPROVED** [review]

**Pull request merged**